### PR TITLE
Clean up version-tools APIs and enable api-extractor

### PR DESCRIPTION
**Pull request by tylerbutler**
Enables API extractor on the version-tools project and cleans up some of the API signatures.

Also refactors how the virtual patch format is handled and adds more tests.

### CODE REVIEW

File: build-tools/packages/version-tools/api-extractor.json [hunk 1]
```diff
@@ -1,4 +1,16 @@
...
-  "extends": "@fluidframework/build-common/api-extractor-common-strict.json"
+  "extends": "@fluidframework/build-common/api-extractor-common-report.json",
+  "apiReport": {
+    "enabled": true,
+    "reportFolder": "<projectFolder>/api-report/"
+  },
+  "messages": {
+    "extractorMessageReporting": {
+      "ae-unresolved-link": {
+        "logLevel": "warning",
```
Comment (Josmithr) [inline build-tools/packages/version-tools/api-extractor.json:11]: Are there currently broken links? Can we not reasonably fix them before merging?

File: build-tools/packages/version-tools/src/semver.ts [hunk 2]
```diff
@@ -89,3 +89,52 @@ export function detectConstraintType(range: string): "minor" | "patch" {
...
+        throw new Error(`Invalid version: ${v1}`);
+    }
+
+    const v2Parsed = semver.parse(v2);
+    if (v2Parsed === null) {
+        throw new Error(`Invalid version: ${v2}`);
+    }
+
+    if (semver.compareBuild(v1Parsed, v2Parsed) >= 0) {
+        return;
```
Comment (Josmithr) [inline build-tools/packages/version-tools/src/semver.ts:116]: Is this a scenario we should guard against / report an error if we see it? This means the first version is semantically greater than the second, right? If returning undefined is the right move here, we might want to note it explicitly in the docs.

I would also recommend the following to be explicit:
```suggestion
        return undefined;
```

File: build-tools/packages/version-tools/src/virtualPatchScheme.ts [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { strict as assert } from "assert";
+import * as semver from "semver";
+import { VersionBumpType } from "./bumpTypes";
+
+/**
+ * Determines if a version is a virtual patch format or not using a very simplistic algorithm.
```
Comment (Josmithr) [inline build-tools/packages/version-tools/src/virtualPatchScheme.ts:11]: Nit
```suggestion
 * Determines if a version is a virtual patch format or not, using a very simplistic algorithm.
```

File: build-tools/packages/version-tools/src/virtualPatchScheme.ts [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+ */
+export function isVirtualPatch(version: semver.SemVer | string): boolean {
+    // If the major is 0 and the patch is >= 1000 assume it's a virtualPatch version
+    if (semver.major(version) === 0 && semver.patch(version) >= 1000) {
```
Comment (Josmithr) [inline build-tools/packages/version-tools/src/virtualPatchScheme.ts:15]: `1000` appears in a lot of places. Should we define a constant for it somewhere and document it?

**Review by Josmithr — APPROVED** [review]

File: build-tools/packages/version-tools/api-extractor.json (hunk 1 above)
Comment (tylerbutler) [reply build-tools/packages/version-tools/api-extractor.json:11]: Yeah, there are currently some types that are exported by the build-tools project that doesn't have api-extractor enabled. I'm leaving this for later; once we're fully moved to the new tools, those types will be moved from the old build-tools project into the CLI project.

**Comment by tylerbutler** [comment]
Thanks for the review, @Josmithr !

**Pull request merged**